### PR TITLE
[rush-lib] handle optional dependencies in transitive injected install

### PR DESCRIPTION
**Pull request by g-chao**
## Summary
This PR adds some improvements to the PR https://github.com/microsoft/rushstack/pull/4583, which relates to PNPM injected install feature for Rush Subspace. 

## Details
When a package gets installed in the injected way then every workspace:* dependency of that package also gets injected.
Here, the transitive injected dependency should only apply to dependencies and optional dependencies fields in the `package.json`, not include the devDependencies

### CODE REVIEW

**Review by iclanton — APPROVED** [review]
What about peerDependencies?

**Comment by g-chao** [comment]
> What about peerDependencies?

Is there any case that use `workspace:*` as a peer dependency version?

**Comment by octogonz** [comment]
> Is there any case that use `workspace:*` as a peer dependency version?

Historically no, because peers wouldn't get installed correctly with `workspace:*`. But isn't `pnpm-sync` the fix for that? 😄

File: common/changes/@microsoft/rush/chao-improve-transitive-injected-install_2024-04-04-22-21.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Handle optional dependencies in transitive injected install",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/chao-improve-transitive-injected-install_2024-04-04-22-21.json:5]:
```suggestion
      "comment": "Add support for `optionalDependencies` in transitive injected install in the Subspaces feature.",
```

**Review by iclanton — APPROVED** [review]

**Comment by octogonz** [comment]
> Historically no, because peers wouldn't get installed correctly with `workspace:*`. But isn't `pnpm-sync` the fix for that? 

I probably should have asked this more clearly and replied as a diff comment that blocks auto-merge. 🙂 

@g-chao  Shouldn't `peerDependencies` be transitive injected as well? I'm not 100% sure about the answer.

Also, regarding `optionalDependencies`, it's interesting to consider how PNPM should handle that.  From what I understand:
- Except for metadata constraints like `os` or `cpu`, PNPM normally downloads the tarball of an optional dependency and attempts to install it; if the `"install"` lifecycle script fails (nonzero exit code) then PNPM aborts that package and can continue to a successful installation.
- This logically normally would not be applied for `workspace:*` dependency, whereas for an injected dependency, the above algorithm seems straightforward since the only difference is where the tarball comes from.
- However prior to `pnpm-sync` that would have been an extremely rare edge case; it might be worth testing to confirm that optional injected dependencies actually get installed correctly (for both `os` constraints and lifecycle script failures).
- `optionalPeerDependencies` are another (probably much more common) edge case; their semantics are entirely different from `optionalDependencies`, but might be worth testing as well just to be sure

**Pull request merged**